File: desktop/src/vault/prefs.ts
// User preferences + theme. The design exposes accent / font / writing-column /
// frontmatter-style as canvas props and theme / font-override / sidebar-width /
// hidden-files / pretty-names as in-app state; here they are one persisted Prefs
// object (localStorage `asp.prefs.v1`). No backend involvement — these are local
// view settings only.

export type FontKey = 'Sans' | 'Serif' | 'Mono';
export type FrontmatterStyle = 'Card' | 'Banner' | 'Below';
export type Theme = 'light' | 'dark';

export interface Prefs {
  accent: string;
  font: FontKey; // base reading font
  fontOverride: FontKey | null; // status-bar toggle (Serif ⇄ Sans)
  frontmatterStyle: FrontmatterStyle;
  writingColumn: boolean;
  theme: Theme;
  sidebarW: number;
  histBarH: number;
  showHidden: boolean;
  prettyNames: boolean;
}

export const DEFAULT_PREFS: Prefs = {
  accent: '#3d63dd',
  font: 'Sans',
  fontOverride: null,
  frontmatterStyle: 'Below',
  writingColumn: true,
  theme: 'light',
  sidebarW: 266,
  histBarH: 150,
  showHidden: false,
  prettyNames: false,
};

export const FONT_FAMILIES: Record<FontKey, string> = {
  Sans: "system-ui, -apple-system, 'Segoe UI', sans-serif",
  Serif: "'Newsreader', Georgia, serif",
  Mono: "'JetBrains Mono', ui-monospace, Menlo, monospace",
};

const KEY = 'asp.prefs.v1';

export function loadPrefs(): Prefs {
  try {
    const raw = localStorage.getItem(KEY);
    if (raw) return { ...DEFAULT_PREFS, ...(JSON.parse(raw) as Partial<Prefs>) };
  } catch {
    /* ignore */
  }
  return { ...DEFAULT_PREFS };
}

export function savePrefs(prefs: Prefs): void {
  try {
    localStorage.setItem(KEY, JSON.stringify(prefs));
  } catch {
    /* ignore */
  }
}

// The whole app themes from the `data-theme` attribute on <html> (styles.css
// reads it). Mirrors the design's `applyTheme`.
export function applyTheme(theme: Theme): void {
  try {
    document.documentElement.setAttribute('data-theme', theme === 'dark' ? 'dark' : 'light');
  } catch {
    /* ignore */
  }
}

// The effective reading font: the status-bar override wins over the base font.
export function fontFamilyOf(prefs: Prefs): string {
  return FONT_FAMILIES[prefs.fontOverride || prefs.font];
}

export const SIDEBAR_MIN = 200;
export const SIDEBAR_MAX = 460;
export function clampSidebar(w: number): number {
  return Math.max(SIDEBAR_MIN, Math.min(SIDEBAR_MAX, w));
}

// The bottom history/log bar grows upward from the bottom. One shared height
// drives whichever panel is open; dragging below COLLAPSE snaps it fully shut.
export const HISTBAR_MIN = 96;
export const HISTBAR_MAX = 640;
export const HISTBAR_COLLAPSE = 72;
export function clampHistBar(h: number): number {
  return Math.max(HISTBAR_MIN, Math.min(HISTBAR_MAX, h));
}

File: gpui/src/vault/prefs.rs
//! User preferences (accent, writing column, theme, sidebar/history sizes,
//! hidden files, pretty names), ported from desktop `src/vault/prefs.ts`. The
//! localStorage/DOM bits are the app layer's job; here we keep the data model,
//! defaults, and the size clamps (pure + tested).

use crate::theme::Appearance;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FrontmatterStyle {
    Card,
    Banner,
    Below,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Prefs {
    /// Accent color as 0xRRGGBB (default 0x3d63dd).
    pub accent: u32,
    pub frontmatter_style: FrontmatterStyle,
    pub writing_column: bool,
    pub theme: Appearance,
    pub sidebar_w: f32,
    pub hist_bar_h: f32,
    pub show_hidden: bool,
    pub pretty_names: bool,
}

impl Default for Prefs {
    fn default() -> Self {
        Prefs {
            accent: 0x3d63dd,
            frontmatter_style: FrontmatterStyle::Below,
            writing_column: true,
            theme: Appearance::Light,
            sidebar_w: 266.0,
            hist_bar_h: 150.0,
            show_hidden: false,
            pretty_names: false,
        }
    }
}

pub const SIDEBAR_MIN: f32 = 200.0;
pub const SIDEBAR_MAX: f32 = 460.0;

pub fn clamp_sidebar(w: f32) -> f32 {
    w.max(SIDEBAR_MIN).min(SIDEBAR_MAX)
}

pub const HISTBAR_MIN: f32 = 96.0;
pub const HISTBAR_MAX: f32 = 640.0;
pub const HISTBAR_COLLAPSE: f32 = 72.0;

pub fn clamp_hist_bar(h: f32) -> f32 {
    h.max(HISTBAR_MIN).min(HISTBAR_MAX)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn defaults_match_design() {
        let p = Prefs::default();
        assert_eq!(p.accent, 0x3d63dd);
        assert_eq!(p.sidebar_w, 266.0);
        assert_eq!(p.hist_bar_h, 150.0);
        assert!(p.writing_column);
        assert!(!p.show_hidden);
        assert_eq!(p.theme, Appearance::Light);
    }

    #[test]
    fn clamp_sidebar_bounds() {
        assert_eq!(clamp_sidebar(10.0), SIDEBAR_MIN);
        assert_eq!(clamp_sidebar(9999.0), SIDEBAR_MAX);
        assert_eq!(clamp_sidebar(300.0), 300.0);
    }

    #[test]
    fn clamp_hist_bar_bounds() {
        assert_eq!(clamp_hist_bar(10.0), HISTBAR_MIN);
        assert_eq!(clamp_hist_bar(9999.0), HISTBAR_MAX);
        assert_eq!(clamp_hist_bar(150.0), 150.0);
    }
}
